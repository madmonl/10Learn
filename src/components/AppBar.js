import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import { Button, IconButton, Typography, Toolbar, AppBar, Tabs, Tab, Badge } from '@material-ui/core/';
import MenuIcon from '@material-ui/icons/Menu';
import { startLogout } from '../actions/auth';
import { connect } from 'react-redux';
import compose from 'recompose/compose';
import styled from 'styled-components';
import { PowerSettingsNew, Folder, Edit } from '@material-ui/icons';

// <div className={classes.container}>
const ToolbarCSS = styled(Toolbar)`
  display: flex;
  justify-content: space-between;
`;

const styles = {
  button: {
    size: 70,
    width: 100,
    fontSize: 15
  }
};

export class ButtonAppBar extends Component {
  constructor(props) {
     super(props);
  }
  logout = () => {
    if (this.props.guest) {
        this.props.dispatchGuestLogout();
    }
    else {
        this.props.startLogout();
    }
}

  render () {
    const { classes } = this.props;
    return (
        <div>
            <AppBar position="static" color="default">
              <div className="app-bar-header__theme">            
                <div className="content-container">
                  <ToolbarCSS>
                    <Typography color="textPrimary" variant="display2">
                        10Learn
                    </Typography>
                    <div className='app-bar-header'>
                      <Tabs value={0}>              
                        <Tab label={<span className="app-bar-header__tab">
                          <Edit />
                          תרגול
                        </span>} />
                        <Tab label={<span className="app-bar-header__tab">
                          <Folder />
                          מבחנים קודמים
                        </span>} />
                      </Tabs>
                      <Button className={classes.button} size='medium' onClick={this.logout}>
                        <PowerSettingsNew />
                        התנתק
                      </Button>
                    </div>          
                  </ToolbarCSS>
                </div>                
              </div>
            </AppBar>
        </div>
    );
  };
}

ButtonAppBar.propTypes = {
  classes: PropTypes.object.isRequired,
};

const mapDispatchToProps = (dispatch) => ({
    startLogout: () => dispatch(startLogout()),
    dispatchGuestLogout: () => dispatch(dispatchGuestLogout())
});

const mapStateToProps = (state) => ({
    guest: state.auth.guest
});

export default compose(
    withStyles(styles), 
    connect(mapStateToProps, mapDispatchToProps)
  )(ButtonAppBar);