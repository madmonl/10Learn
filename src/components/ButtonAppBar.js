import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import { Button, IconButton, Typography, Toolbar, AppBar, Tabs, Tab, Badge, Paper } from '@material-ui/core/';
import MenuIcon from '@material-ui/icons/Menu';
import { startLogout, dispatchGuestLogout } from '../actions/auth';
import { connect } from 'react-redux';
import compose from 'recompose/compose';
import styled from 'styled-components';
import { PowerSettingsNew, Folder, Edit } from '@material-ui/icons';
import { dispatchHeaderTabsNavigation } from '../actions/material'

// <div className={classes.container}>
const ToolbarCSS = styled(Toolbar)`
  display: flex;
  justify-content: space-between;
`;

const TypographyTABS = styled(Typography)`
  display: flex;
  justify-content: center;
  margin-right: 20px;
`;

const TypographyCSS = styled(Typography)`
  font: italic bold Georgia;
`;

const styles = theme => ({
  button: {
    size: 70,
    width: 100,
    fontSize: 15
  }
});

export class ButtonAppBar extends Component {
  constructor(props) {
    super(props);
    this.state = {
      appBarTabsIndex: 0
    }
    this.logout = this.logout.bind(this);
    this.onIndexSelect = this.onIndexSelect.bind(this);
  }
  
  logout = () => {
    if (this.props.guest) {
        this.props.dispatchGuestLogout();
    }
    else {
        this.props.startLogout();
    }
  }

  onIndexSelect = (e, index) => {
    this.props.dispatchHeaderTabsNavigation(this.props.tabs[index]);
  }

  render () {
    const { classes, tabs, currTab } = this.props;
    const currTabIndex = tabs.indexOf(currTab);
    return (
        <div>
            <AppBar position="static" color="default">
              <div className="app-bar-header__theme">            
                <div className="content-container">
                  <ToolbarCSS>
                    <TypographyCSS color="textPrimary" variant="display2">
                        10Learn
                    </TypographyCSS>
                    <div className='app-bar-header'>
                      <Tabs 
                        value={currTabIndex}
                        indicatorColor="secondary"
                        onChange={this.onIndexSelect}
                      >
                        <Tab label={<span className="app-bar-header__tab">
                          <Edit className="app-bar-header__icon" />
                          <TypographyTABS className="app-bar-header__text" color="textPrimary" variant="title">
                            תרגול
                          </TypographyTABS>
                        </span>} />
                        <Tab label={<span className="app-bar-header__tab">
                          <Folder className="app-bar-header__icon" />
                          <TypographyTABS className="app-bar-header__text" color="textPrimary" variant="title">
                            מבחנים קודמים
                          </TypographyTABS>
                        </span>} />
                      </Tabs>
                      <Button className={classes.button} size='medium' onClick={this.logout}>
                        <PowerSettingsNew />
                        התנתק
                      </Button>
                    </div>       
                  </ToolbarCSS>
                </div>                
              </div>
            </AppBar>
        </div>
    );
  };
}

const mapDispatchToProps = (dispatch) => ({
    startLogout: () => dispatch(startLogout()),
    dispatchGuestLogout: () => dispatch(dispatchGuestLogout()),
    dispatchHeaderTabsNavigation: (currTab) => dispatch(dispatchHeaderTabsNavigation(currTab))
});

const mapStateToProps = (state) => ({
    guest: state.auth.guest,
    currTab: state.material.currTab,
    tabs: state.material.tabs
});

export default compose(
    withStyles(styles), 
    connect(mapStateToProps, mapDispatchToProps)
  )(ButtonAppBar);