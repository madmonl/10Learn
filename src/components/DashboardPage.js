import React from 'react';

const DashboardPage = () => (
    <div>
        10Learn
    </div>
);

export default DashboardPage;
