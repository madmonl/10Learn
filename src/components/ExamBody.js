import React, { Fragment, Component } from 'react';
import MathJax from 'react-mathjax';
import { dispatchChangeQuestion, dispatchMarkQuestion, dispatchSetAnswer } from '../actions/exam';
import { connect } from 'react-redux';

export class ExamBody extends Component {
  constructor (props) {
    super(props);
    this.solutionClassName = this.solutionClassName.bind(this);
    this.onSubmitQuestion = this.onSubmitQuestion.bind(this);
  }

  onSubmitQuestion = (index, answer) => {
    this.props.dispatchChangeQuestion((index + 1) % 10);
    this.props.dispatchMarkQuestion(index);
    this.props.dispatchSetAnswer(index, answer);
  } 
  
  solutionClassName = (index, questions, currQuestion, questionsStatus, answersStatus) => {
    // one of:
    // "exam__answer-item" when still in exam - clickable.
    // "exam__answer-correct" not clickable - correct solution only.
    // "exam__answer-mistake" only a mistaken solution that was chosen.
    // or exam finished and
    // "exam__answer-default" not clickable not chosen, not correct answer.
    if (questionsStatus[currQuestion] === 'being_answered') {
      return "exam__answer-item"
    } else if (questions[currQuestion].index === index) {
      return "exam__answer-correct exam__answer-item"
    } else if (questions[currQuestion].index !== index 
      && answersStatus[currQuestion] === index) {
      return "exam__answer-mistake exam__answer-item"
    } else {
      return "exam__answer-default exam__answer-item"
    };
  }
  
  render () {
    const { questions, currQuestion, questionsStatus, answersStatus } = this.props;
    
    return (
      <div>
        <div className="exam__question-navigation-container">
          <div className="exam__question">
            <MathJax.Provider>
              <div>
                  <MathJax.Node formula={questions[currQuestion].question} />
              </div>
            </MathJax.Provider>
          </div>
        </div>
        <div className="exam__answers">
          {questions[currQuestion].solutions.map((solution, index) =>
            <div 
              className={
                this.solutionClassName(index, questions, currQuestion, questionsStatus, answersStatus)
              }
              key={index}
              onClick={() => this.onSubmitQuestion(currQuestion, index)}
            >
              <MathJax.Provider>
                <div>
                    <MathJax.Node formula={solution} />
                </div>
              </MathJax.Provider>
            </div>             
          )}
        </div>
      </div>
    )
  }
}

const mapDispatchToProps = (dispatch) => ({
    dispatchChangeQuestion: (question) => dispatch(dispatchChangeQuestion(question)),
    dispatchMarkQuestion: (index) => dispatch(dispatchMarkQuestion(index)),
    dispatchSetAnswer: (index, answer) => dispatch(dispatchSetAnswer(index, answer))
});

const mapStateToProps = (state) => ({
    questions: state.exam.questions,
    currQuestion: state.exam.currQuestion,
    questionsStatus: state.exam.questionsStatus,
    answersStatus: state.exam.answersStatus,
});
  
export default connect(mapStateToProps, mapDispatchToProps)(ExamBody);
