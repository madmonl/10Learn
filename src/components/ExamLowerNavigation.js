
import React, { Fragment, Component } from 'react';
import { dispatchChangeQuestion } from '../actions/exam';
import { Button } from '@material-ui/core/';
import { withStyles } from '@material-ui/core/styles';
import { Send } from '@material-ui/icons';
import { connect } from 'react-redux';
import compose from 'recompose/compose';

export const styles = theme => ({
  button: {
    margin: theme.spacing.unit
  },
  rightIcon: {
    marginLeft: theme.spacing.unit
  },
  leftIcon: {
    marginLeft: theme.spacing.unit
  }
})

export class ExamLowerNavigation extends Component {
  constructor (props) {
    super(props);
    this.onQuestionChange = this.onQuestionChange.bind(this);
  }
  
  onQuestionChange = direction => {
    if (direction === 'prev'){
      this.props.dispatchChangeQuestion( this.props.currQuestion === 0 
        ? this.props.questions.length - 1
        : this.props.currQuestion - 1)
    } else if (direction === 'next') {
      this.props.dispatchChangeQuestion((this.props.currQuestion + 1) % this.props.questions.length)
    } else {
      throw "Button direction is incorrect.";
    }
  }

  render () {
    const { questions, currQuestion, classes } = this.props;
    
    return (
      <div>
        <div className="exam__navigation-container">
          <Button 
            variant="contained" 
            color="primary" 
            className={classes.button}
            onClick={() => this.onQuestionChange('prev')}
          >
            <Send className={classes.rightIcon} />
            שאלה קודמת            
          </Button>
          <Button 
            variant="contained" 
            color="primary" 
            className={classes.button}
            onClick={() => this.onQuestionChange('next')}
          >
            שאלה הבאה
            <Send className={classes.leftIcon} />
          </Button>
        </div>
      </div>
    )
  }
}

const mapDispatchToProps = (dispatch) => ({
    dispatchChangeQuestion: (question) => dispatch(dispatchChangeQuestion(question))
});

const mapStateToProps = (state) => ({
    questions: state.exam.questions,
    currQuestion: state.exam.currQuestion
});
  
export default compose(
  withStyles(styles), 
  connect(mapStateToProps, mapDispatchToProps))
  (ExamLowerNavigation);

    