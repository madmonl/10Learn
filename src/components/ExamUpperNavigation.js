import React, { Component } from 'react';
import NumbersNavigation from './NumbersNavigation';
import SubmitExamModal from './SubmitExamModal';




export default () =>
  <div className="exam__upper-footer">
    <NumbersNavigation />
    <SubmitExamModal />
  </div>

