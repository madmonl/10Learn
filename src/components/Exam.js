import React, { Fragment, Component } from 'react';
import ExamUpperNavigation from './ExamUpperNavigation';
import ExamBody from './ExamBody';
import ExamLowerNavigation from './ExamLowerNavigation';

export default () => 
  <div className="exam">
    <ExamUpperNavigation />
    <ExamBody />
    <ExamLowerNavigation />        
  </div>

