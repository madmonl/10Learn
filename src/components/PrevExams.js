import React, { Fragment, Component } from 'react';
import { Typography, Paper, Modal, Slide } from '@material-ui/core/';
import { Chip } from '@material-ui/core/';
import { withStyles } from '@material-ui/core/styles';
import { connect } from 'react-redux';
import compose from 'recompose/compose';
import NumbersNavigation from './NumbersNavigation';
import Exam from './Exam';
import { dispatchChangeExam } from '../actions/exam'

const styles = theme => ({
  paper: { 
    padding: 0, 
    overflowY: 'auto',
    height: '900px',
    marginTop: '5px'
  },
  chip: {
    margin: theme.spacing.unit,
    width: 40
  },
});

function getModalStyle() {
  return {
    margin:'auto'
  };
}

export class PrevExams extends Component {
  constructor (props) {
    super(props);
    this.state = {
      open: false,
      modalExam: {}
    }
    this.changeExamModal = this.changeExamModal.bind(this);
  }

  changeExamModal = exam => {
    this.setState({ open: true, modalExam: exam });
    // needed because <Exam /> is being rendered and it's values are taken from the store.
    this.props.dispatchChangeExam(exam);
  }
  render () {
    const { classes, prevExams, uid } = this.props,
          { open, modalExam } = this.state
    
    return (
      <div className="content-container">
        <Paper className={classes.paper}>
        {          
          uid 
          ? <div className="prevExams">
              {prevExams.map((exam) => 
                <div
                  className="prevExam"
                  key={exam.id}
                  onClick={() => this.changeExamModal(exam)}
                >
                  <div>
                    {exam.selectedSubjects.map((subject, index) =>
                      <Chip 
                        key={index}
                        label={<span className="chipLabel">{subject}</span>}
                        className={classes.chip}
                        color="primary"
                        variant="outlined"
                      />
                    )}
                  </div>
                  <NumbersNavigation               
                    questions={exam.questions}
                    currQuestion={0}
                    answeredQuestions={[]}
                    questionsStatus={exam.questionsStatus}  
                  />                
                </div>            
              )}
              <Modal
                aria-labelledby="simple-modal-title"
                aria-describedby="simple-modal-description"
                open={open}
                onClose={() => this.setState({ open: false })}
                className="MuiModal-root-container"
              >
                <Slide direction="up" in={open} mountOnEnter unmountOnExit>
                  <div style={getModalStyle()} className="exam__modal-box">
                    <Exam />
                  </div>
                </Slide>
              </Modal>      
            </div>                             
          : <p>התחבר/י בשביל לראות מבחנים קודמים</p>
        }
        </Paper> 
      </div>
    );
  };
};

const mapDispatchToProps = dispatch => ({
  dispatchChangeExam: (exam) => dispatch(dispatchChangeExam(exam))
});

const mapStateToProps = (state) => ({
  prevExams: state.exam.prevExams,
  uid: state.auth.uid
})

export default compose(
    withStyles(styles), 
    connect(mapStateToProps, mapDispatchToProps)
  )(PrevExams);