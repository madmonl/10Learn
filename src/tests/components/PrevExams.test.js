import React from 'react';
import { shallow } from 'enzyme';
import { PrevExams } from '../../components/PrevExams';
import exams from '../fixtures/exams';
import { Chip } from '@material-ui/core/';


let wrapper, classes, dispatchChangeExam;

beforeEach(() => {
  dispatchChangeExam = jest.fn();
  wrapper = shallow(
    <PrevExams 
      dispatchChangeExam={dispatchChangeExam}
      classes={{}}
      prevExams={exams}
    />
  );
});

test('should render PrevExams correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should open the first exam in the modal', () => {
  wrapper.find('.prevExams').children().at(0).prop('onClick')();
  expect(wrapper.state('open')).toBe(true);
  expect(wrapper.state('modalExam')).toEqual(exams[0]);
  expect(dispatchChangeExam).toHaveBeenCalledWith(exams[0])
});

test('rendered Chip should match the selectedSubjects', () => {
  expect(wrapper.find(Chip).length).toBe(
    exams[0].selectedSubjects.length + exams[1].selectedSubjects.length)
});