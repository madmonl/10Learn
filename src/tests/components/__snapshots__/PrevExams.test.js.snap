// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render PrevExams correctly 1`] = `
<div
  className="content-container"
>
  <WithStyles(Paper)>
    <div
      className="prevExams"
    >
      <div
        className="prevExam"
        key="1"
        onClick={[Function]}
      >
        <div>
          <WithStyles(Chip)
            color="primary"
            key="0"
            label="חיבור"
            variant="outlined"
          />
          <WithStyles(Chip)
            color="primary"
            key="1"
            label="חיסור"
            variant="outlined"
          />
        </div>
        <WithStyles(Connect(NumbersNavigation))
          answeredQuestions={Array []}
          currQuestion={0}
          questions={
            Array [
              Object {
                "index": 1,
                "question": "1 + 2 = ?",
                "solutions": Array [
                  1,
                  3,
                  4,
                ],
              },
              Object {
                "index": 1,
                "question": "1 + 2 = ?",
                "solutions": Array [
                  1,
                  3,
                  4,
                ],
              },
              Object {
                "index": 1,
                "question": "1 + 2 = ?",
                "solutions": Array [
                  1,
                  3,
                  4,
                ],
              },
              Object {
                "index": 1,
                "question": "1 + 2 = ?",
                "solutions": Array [
                  1,
                  3,
                  4,
                ],
              },
              Object {
                "index": 1,
                "question": "1 + 2 = ?",
                "solutions": Array [
                  1,
                  3,
                  4,
                ],
              },
              Object {
                "index": 1,
                "question": "1 + 2 = ?",
                "solutions": Array [
                  1,
                  3,
                  4,
                ],
              },
              Object {
                "index": 1,
                "question": "1 + 2 = ?",
                "solutions": Array [
                  1,
                  3,
                  4,
                ],
              },
              Object {
                "index": 1,
                "question": "1 + 2 = ?",
                "solutions": Array [
                  1,
                  3,
                  4,
                ],
              },
              Object {
                "index": 1,
                "question": "1 + 2 = ?",
                "solutions": Array [
                  1,
                  3,
                  4,
                ],
              },
              Object {
                "index": 1,
                "question": "1 + 2 = ?",
                "solutions": Array [
                  1,
                  3,
                  4,
                ],
              },
            ]
          }
          questionsStatus={
            Array [
              "mistake",
              "mistake",
              "mistake",
              "mistake",
              "mistake",
              "mistake",
              "mistake",
              "mistake",
              "mistake",
              "mistake",
            ]
          }
        />
      </div>
      <div
        className="prevExam"
        key="2"
        onClick={[Function]}
      >
        <div>
          <WithStyles(Chip)
            color="primary"
            key="0"
            label="כפל"
            variant="outlined"
          />
          <WithStyles(Chip)
            color="primary"
            key="1"
            label="חילוק"
            variant="outlined"
          />
        </div>
        <WithStyles(Connect(NumbersNavigation))
          answeredQuestions={Array []}
          currQuestion={0}
          questions={
            Array [
              Object {
                "index": 2,
                "question": "4 + 3 = ?",
                "solutions": Array [
                  5,
                  6,
                  7,
                ],
              },
            ]
          }
          questionsStatus={
            Array [
              "mistake",
              "mistake",
              "mistake",
              "mistake",
              "mistake",
              "mistake",
              "mistake",
              "mistake",
              "mistake",
              "mistake",
            ]
          }
        />
      </div>
      <WithStyles(Modal)
        aria-describedby="simple-modal-description"
        aria-labelledby="simple-modal-title"
        className="MuiModal-root-container"
        onClose={[Function]}
        open={false}
      >
        <WithTheme(Slide)
          direction="up"
          in={false}
          mountOnEnter={true}
          unmountOnExit={true}
        >
          <div
            className="exam__modal-box"
            style={
              Object {
                "margin": "auto",
              }
            }
          >
            <Component />
          </div>
        </WithTheme(Slide)>
      </WithStyles(Modal)>
    </div>
  </WithStyles(Paper)>
</div>
`;
